### PR TITLE
Drop `parameterized_class` from `test_image`

### LINKED ISSUE #803: Drop `parameterized_class`
We should drop `@parameterized_class` going forward in favor of just using `pytest` directly.

To remove `parameterized_class` from our package would require addressing these locations.

```
tests/test_FBbasis2D.py:from parameterized import parameterized_class
tests/test_FBbasis2D.py:@parameterized_class(
tests/test_FFBbasis2D.py:from parameterized import parameterized_class
tests/test_FFBbasis2D.py:@parameterized_class(
tests/test_image.py:from parameterized import parameterized_class
tests/test_image.py:@parameterized_class(("parity",), [(0,), (1,)])
tests/test_synthetic_volume.py:from parameterized import parameterized_class
tests/test_synthetic_volume.py:@parameterized_class(
tests/test_synthetic_volume.py:@parameterized_class(
tests/test_synthetic_volume.py:@parameterized_class(("L"), [(21,)])
tests/test_synthetic_volume.py:@parameterized_class(("L"), [(21,)])
tests/test_synthetic_volume.py:@parameterized_class(("L"), [(21,)])
tests/test_synthetic_volume.py:@parameterized_class(("L"), [(21,)])
```

### PR DESCRIPTION
**Pull request by chris-langfield**
WIP to close #803 

In adding parametrization over dtype, I have found an issue with `Image` shifting for `np.float32`.

### CODE REVIEW

**Comment by chris-langfield** [comment]
It turned out that passing the real image data as `float32` into the FFT returns a different value from passing it in as `float64`, which caused numerical instability in the rest of the shifting algorithm. This PR patches the `_im_translate` method by upcasting the real data before the shifting algorithm, and then casting back to the original dtype before return.

File: tests/test_image.py [hunk 1]
```diff
@@ -1,233 +1,282 @@
...
-    def testMultiDimBroadcast(self):
-        X = self.mdim_ims + self.ims
-        self.assertTrue(np.allclose(X[0], 2 * self.ims.asnumpy()))
+n = 3
+mdim = 2
+
+
+def get_images(parity=0, dtype=np.float32):
+    size = 768 - parity
+    im_np = misc.face(gray=True).astype(dtype)
```
Comment (garrettwrong) [inline tests/test_image.py]: redundant, delete

File: tests/test_image.py [hunk 2]
```diff
@@ -1,233 +1,282 @@
...
+    # numpy array for top-level functions that directly expect it
+    im_np = misc.face(gray=True).astype(dtype)[np.newaxis, :size, :size]
```
Comment (garrettwrong) [inline tests/test_image.py:25]: lets normalize these values to 0 1
```
def get_images(parity=0, dtype=np.float32):
    size = 768 - parity
    # numpy array for top-level functions that directly expect it                                                                
    im_np = misc.face(gray=True).astype(dtype)[np.newaxis, :size, :size]
    denom = np.max(np.abs(im_np))
    im_np = im_np / denom  # Normalize test image data to 0,1                                                                    
    # Independent Image object for testing Image methods                                                                         
    im = Image(im_np.copy())
    return im_np, im
```

File: src/aspire/image/image.py [hunk 3]
```diff
@@ -341,7 +341,8 @@ def _im_translate(self, shifts):
 
         # Note original stack shape and flatten stack
         stack_shape = self.stack_shape
-        im = self.stack_reshape(-1)._data
+        # upcast to np.float64 to preserve numerical stability in FFT ops
+        im = self.stack_reshape(-1)._data.astype(np.float64)
```
Comment (garrettwrong) [inline src/aspire/image/image.py]: see my other remarks. This file can be reverted to `origin/develop` if you take that other patch.

File: tests/test_image.py [hunk 4]
```diff
@@ -1,233 +1,282 @@
...
+    # Multi dimensional stack Image object
+    mdim = 2
+    mdim_ims_np = np.concatenate([ims_np] * mdim).reshape(mdim, *ims_np.shape)
+    mdim_ims = Image(mdim_ims_np)
+
+    return mdim_ims_np, mdim_ims
+
+
+def testRepr():
+    # don't need to parametrize this test
```
Comment (garrettwrong) [inline tests/test_image.py]: I think we can rm the `# don't need to parametrize this test`, or at least, put it under the docstring.

**Review by garrettwrong — CHANGES_REQUESTED** [review]
I took a look at the issue visually and numerically. The relative error seemed okay to me, so I am suggesting normalizing the image data before the test instead of bumping the `atol` or casting.  That appears to work well.  

The input data here is [0,250].  Another thing that I can think of is that we might have a normalization bug somewhere in the algo.

This makes me wonder why we don't just use roll for integer shifts 😇  and reserve this for subpixel or shifts already in Fourier space. I'll make an issue for that.

**Comment by garrettwrong** [comment]
Created #822

File: tests/test_image.py (hunk 1 above)
Comment (chris-langfield) [reply tests/test_image.py]: done thanks

File: tests/test_image.py (hunk 2 above)
Comment (chris-langfield) [reply tests/test_image.py:25]: cool, thank you for looking into this!

**Review by garrettwrong — DISMISSED** [review]
Cool, thanks for making the changes. Let's get one more pair of eyes on it (Josh) before merging.

File: tests/test_image.py [hunk 5]
```diff
@@ -1,233 +1,282 @@
...
+n = 3
+mdim = 2
+
+
+def get_images(parity=0, dtype=np.float32):
+    size = 768 - parity
+    # numpy array for top-level functions that directly expect it
+    im_np = misc.face(gray=True).astype(dtype)[np.newaxis, :size, :size]
+    denom = np.max(np.abs(im_np))
+    im_np = im_np / denom  # Normalize test image data to 0,1
```
Comment (j-c-c) [inline tests/test_image.py]: `im_np = im_np / denom ` --> `im_np /= denom`

File: tests/test_image.py [hunk 6]
```diff
@@ -1,233 +1,282 @@
...
+
+    return ims_np, ims
+
+
+def get_mdim_images(parity=0, dtype=np.float32):
+    ims_np, im = get_stacks(parity, dtype)
+    # Multi dimensional stack Image object
+    mdim = 2
+    mdim_ims_np = np.concatenate([ims_np] * mdim).reshape(mdim, *ims_np.shape)
+    mdim_ims = Image(mdim_ims_np)
```
Comment (j-c-c) [inline tests/test_image.py]: Maybe add same comment as above: "# Independent multi-dimensional Image stack object for testing Image methods"

File: tests/test_image.py [hunk 7]
```diff
@@ -1,233 +1,282 @@
...
+
+
+def testNonSquare():
+    # don't need to parametrize this test
+    """Test that an irregular Image array raises."""
+    with raises(ValueError, match=r".* square .*"):
+        _ = Image(np.empty((4, 5)))
+
+
+@pytest.mark.parametrize("parity,dtype", params)
```
Comment (j-c-c) [inline tests/test_image.py:71]: No space after the comma in `"parity,dtype"` really bugs me, but I see that's the convention in pytest docs.

**Review by j-c-c — COMMENTED** [review]
Looks good! Couple small things.

**Review by j-c-c — APPROVED** [review]

**Pull request merged**